src/framework/components/element/pointer-events.js

```diff
@@ -0,0 +1,190 @@
+            for(var i = 0; i < this.entity.children.length; i++) {
+                var element =  this.entity.children[i];
+                if (element && element.element) {
+                    element.element[ name ].apply( element.element, eventData );
+                }
+            }
+        },
+
+        // Handles "down" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventDown: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventDown", [ point ]);
+            this.fire(pc.POINTEREVENT_DOWN, point);
+        },
+
+        // Handles "up" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventUp: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventUp", [ point ]);
+            this.fire(pc.POINTEREVENT_CLICK, point);
+            this.fire(pc.POINTEREVENT_UP, point);
+        },
+
+        // Fires pointer leave event and also makes all children do so.
+        _ensurePointerLeaveEvent: function(point) {
+            this._pointerOver = false;
+            this.fire(pc.POINTEREVENT_LEAVE, point); 
+
+            this._passPointerEventToChildren( "_ensurePointerLeaveEvent", [ point ]);
+        },
+
+        // Handles "move" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventMove: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                if (this._pointerOver) {
+                    this._ensurePointerLeaveEvent(point);            
+                }
+
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventMove", [ point ]);
+
+            if (!this._pointerOver) {
+                this._pointerOver = true;
+                this.fire(pc.POINTEREVENT_ENTER, point);                
+            }
+
+            this.fire(pc.POINTEREVENT_MOVE, point);
+        },
+
+        // Handles "scroll" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventScroll: function(point, amount) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventScroll", [ point, amount ]);
+            this.fire(pc.POINTEREVENT_SCROLL, point, amount);
+        },
+
+        // Mouse-specific event handler.
+        _onMouseDown: function(mouseEvent) {
+            this._pointerEventDown( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
```

Comment:
We have had internal conversations about similar cases, where there are events and Vec is returned, what to do here. And we came up to conclusion that as GC is very major issue in pretty much any project we've worked on, it drives some design decisions, and one of them - to make either Vec living for only duration of that callback sync scope, or to provide arguments with X and Y numbers instead of object.

Plus, there is already mouseEvent, why wont we just pass it through?